src/strategies/properties/attribute.sml

```diff
@@ -0,0 +1,106 @@
+import "util.parser";
+import "util.multiset";
+import "util.type";
+
+signature ATTRIBUTE =
+sig
+  structure M : MULTISET;
+  type T;
+  val fromType : Type.T -> T;
+  val fromHoles : Type.T M.multiset -> T;
+  val fromContent : Type.T -> T;
+  val fromFeature : string -> T;
+  val read : string -> T list;
+  val toString : T -> string;
+  val getType : T list -> Type.T option
+end
+
+
+structure Attribute : ATTRIBUTE =
+struct
+
+  structure M = Multiset(struct
+                         type t = Type.T;
+                         val compare = Type.compare;
+                         val fmt = Type.toString;
+                         end);
+
+  datatype T = IsOfType of Type.T
+             | Holes of Type.T M.multiset
+             | Tokens of string list
+             | Content of Type.T
+             | Occurrences of int
+             | Feature of string;
+
+  fun fromType t = IsOfType t;
+  fun fromHoles H = Holes H;
+  fun fromTokens L = Tokens L;
+  fun fromContent c = Content c;
+  fun fromFeature s = Feature s;
+
+  fun decomposeAttribute a =
+      let val (x,_,y) = (Parser.breakOn ":=" a);
+      in (Parser.stripSpaces x,  Parser.stripSpaces y)
+      end;
+
+  fun intFromString s =
+      case Int.fromString s of
+              SOME n => n
+            | NONE => if s = "#t" then ~1 else if s = "sqrt(#t)" then ~2 else if s = "log(#t)" then ~3
```

Comment:
hacking my way into dealing with arithmetic expressions rather than numbers. I expect to change that in the future, but for the moment some fixed expressions are being represented by negative values.
The idea is that some patterns don't have a fixed number of holes, but the number is something like "of the order of the number of tokens" (which would mean that the pattern involves the whole representation [or half, or a fixed fraction]).

Reply:
Ah, that explains what the `#t` is all about then. Perhaps a comment above it explaining what the `#t` is, and that it is to handle some simple algebraic relations.